### ISSUE TITLE
Excel warns before opening XLS files

### ISSUE THREAD

**Issue by simonberger** [highcharts/highcharts#7977]
The in version 5 integrated CSV/XLS exporting module writes insecure XLS files that excel on windows and os x does not open without a warning like described in a former bug (#7064). The problem can't be solved easily because microsoft does not open those html xls files without a warning anymore. Content type does not matter.
Microsoft changed that on all later versions of Excel more than a year ago.

The only solution out of my view would be to integrate a lightweight xls writer plugin. 

#### Live demo with steps to reproduce
http://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/export-data/categorized/

Choose Download XLS and open with Excel

#### Product version
5 and 6 (newest)

#### Affected browser(s)
chrome / firefox

**Comment by TorsteinHonsi** [comment]
See http://jsfiddle.net/highcharts/hr7jLwda/14/ for an example of using third party libraries to support XLS and XLSX export.

We should have a more lightweight solution with less dependencies.

**Comment by simonberger** [comment]
I completely understand this dilemma.
The library demonstrated in your example is one of those I absolutely excluded as it needs more than 500kb summarized which is a no go for this feature.

**Comment by TorsteinHonsi** [comment]
Any suggestion for a lighter library we can plug in?

**Comment by simonberger** [comment]
I just stumbled over https://github.com/egeriis/zipcelx
It is rather new and I didn't try it out yet.

**Comment by simonberger** [comment]
Works quiet good.

Proof of concept.
http://jsfiddle.net/xux1kgvs/

Seems to have some filetype detection problems on iOS at least though

**Comment by TorsteinHonsi** [comment]
Nice, thanks! And only 44kB. 

For the records, if we add a built-in hook for this we need to make sure the licensing is clearly visible. In your built file, there is no license information, plus it seems to import _FileSaver.js_ among others.

**Comment by simonberger** [comment]
@TorsteinHonsi 
I just used the distribution file and did not try to build it myself. But they seem to change much in the building process at the moment. I guess they are happy for all input.

You most likely saw it already. In opposite to my test the result of getDataRows has to be used and converted to the json-config. I would have needed to dive too deep for the test to understand if there are side effects. Especially regarding licensing.

**Comment by simonberger** [comment]
Another package with a similar size and dependencies I had found.
https://github.com/protobi/msexcel-builder

**Comment by andi1984** [comment]
What is the status of this issue? We, the company I'm working for, is planning to roll out XLS download for three different chart types, but this issue makes it hard to go live with xls download into the public with a lot of external customers.

So I would love to see or raise some awareness/need to fix this issue. We would be very pleased if a fix for this could be planned into your release cycle of Highcharts.

_Thanks in advance_

**Comment by TorsteinHonsi** [comment]
We've taken this one step further and provided a recommendation for how to integrate zipcelx. See http://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/export-data/xlsx/. The zipcelx library is MIT licensed, lightweight and standalone.

A possible next step, if this proves stable, would be to integrate a hook that looks for the existence of this library in the page, so you don't need to add the plugin, just the zipcelx file.

**Comment by bdallman** [comment]
Hi @TorsteinHonsi @simonberger 

I've been trying to implement the third party xls to xlsx converter but I've had no luck so far. 

I define the export menu items myself and I'm wondering if this is where I'm running into the issue. Here's a jsfiddle with my menu items defined further down in the JS in your previous example. I'm not sure what I'm currently missing.

http://jsfiddle.net/fd8moqsb/1/     

Any help is greatly appreciated!

**Comment by TorsteinHonsi** [comment]
Yes, your menu item ran `downloadXLS` while it should run `downloadXLSX`: http://jsfiddle.net/highcharts/fd8moqsb/4/

**Comment by bdallman** [comment]
@TorsteinHonsi

**Comment by bdallman** [comment]
@TorsteinHonsi

**Comment by bdallman** [comment]
@TorsteinHonsi

**Comment by pawelfus** [comment]
It looks like probem with zipcelx, see discussion: https://github.com/egeriis/zipcelx/pull/45 and the issue: https://github.com/egeriis/zipcelx/issues/41 
tl;dr: you need to transpile zipcelx to IE11 compatible package, don't link standalone package: `<script src="https://rawgit.com/egeriis/zipcelx/master/lib/standalone.js"></script>
`

**Comment by simonberger** [comment]
I revisit my open bug. As it is nice to see there is a official supported solution with just a bit work to do of one single.
I suggest to switch  away of zipcelx. It is badly maintained right from the start and complicated to implement. I added another possible package https://github.com/protobi/msexcel-builder which looks more promising but I did not test it myself. Maybe there are even better packages available now.

**Comment by chris-vecchio** [comment]
I've created [a plugin](https://github.com/chris-vecchio/export-xlsx) that uses SheetJS's [js-xlsx](https://github.com/SheetJS/js-xlsx) to add basic XLSX export capabilities, and [Moment.js](http://momentjs.com/) for date formatting. See demo [here](https://jsfiddle.net/chrisvecchio/f2r7jqhz/).

I'm a economic analyst/python/Stata/R programmer in my day job so I apologize for inefficient code, lack of tests, etc. I used Moment.js for certain date operations because its api is similar to a Python date library I am familiar with. I imagine removing that dependency would be a positive thing?

I "tested" it using Firefox 66 Dev Edition, Firefox 65, Internet Explorer 11, and Google Chrome 68.

At the recommendation of @pawelfus, I'm considering adding it to the official plugins list.

Hope this helps at least get some ideas going for official  Highcharts XLSX export support. :)

**Comment by simonberger** [comment]
Hi @chris-vecchio  very good initiative !
You used the library one stumbles over first after searching for a xlsx JS library. @TorsteinHonsi mentioned it also at the beginning. But usually the xlsx export is one of many functionalities on a page and 300kb - 900kb (packed to unpacked) for this feature is massive! xlsx brings so many features. Most notable xlsx reading we do not need here.
